File: problems/0210-course-schedule-ii/typescript/solution.ts
// Course Schedule II — https://leetcode.com/problems/course-schedule-ii/
export function findOrder(numCourses: number, prerequisites: number[][]): number[] {
  const graph: number[][] = Array.from({ length: numCourses }, () => []);
  const indegree = new Array<number>(numCourses).fill(0);
  for (const [course, pre] of prerequisites) {
    graph[pre].push(course); // pre -> course
    indegree[course]++;
  }

  const ready: number[] = [];
  for (let c = 0; c < numCourses; c++) {
    if (indegree[c] === 0) ready.push(c);
  }

  const order: number[] = [];
  while (ready.length > 0) {
    ready.sort((a, b) => a - b); // smallest-available-first for determinism
    const c = ready.shift();
    if (c === undefined) break;
    order.push(c);
    for (const nxt of graph[c]) {
      indegree[nxt]--;
      if (indegree[nxt] === 0) ready.push(nxt);
    }
  }

  return order.length === numCourses ? order : [];
}

File: problems/0210-course-schedule-ii/go/solution.go
package solution

import "sort"

// Course Schedule II — https://leetcode.com/problems/course-schedule-ii/
func findOrder(numCourses int, prerequisites [][]int) []int {
	graph := make([][]int, numCourses)
	indegree := make([]int, numCourses)
	for _, p := range prerequisites {
		graph[p[1]] = append(graph[p[1]], p[0]) // pre -> course
		indegree[p[0]]++
	}

	ready := []int{}
	for c := 0; c < numCourses; c++ {
		if indegree[c] == 0 {
			ready = append(ready, c)
		}
	}

	order := []int{}
	for len(ready) > 0 {
		sort.Ints(ready) // smallest-available-first for determinism
		c := ready[0]
		ready = ready[1:]
		order = append(order, c)
		for _, nxt := range graph[c] {
			indegree[nxt]--
			if indegree[nxt] == 0 {
				ready = append(ready, nxt)
			}
		}
	}

	if len(order) == numCourses {
		return order
	}
	return []int{}
}
